tensorflow/c/experimental/gradients/grad_test_helper.cc

```diff
@@ -74,6 +74,39 @@ void CompareNumericalAndAutodiffGradients(
   }
 }
 
+void CompareManualAndAutodiffGradients(
```

Comment:
Can we make this more strict? Right now `manuals = {1.0}` would match both `outputs = {nullptr, 1.0}` and `outputs = {1.0, nullptr}`

The bulk of the logic here is to check the numerical value of the tensor so we could just have a helper for that? Something like

`void CheckTensorValue(const AbstractTensorHandle* output, absl::Span<const int64_T> shape, absl::Span<const float> flat_values)`

Reply:
> Right now manuals = {1.0} would match both outputs = {nullptr, 1.0} and outputs = {1.0, nullptr}

I add a `dims` parameter to avoid confustion.

Reply:
Sorry for nitpicking but this does not look very readable. What do you think about something like:

```
// grad_test_helper.cc

void CheckTensorValue(
    AbstractTensorHandle* t,
    absl::Span<float const> manuals,
    absl::Span<int64_t const> dims,
    double abs_error) {
  TF_Tensor* analytical_tensor;
  s = GetValue(t, &analytical_tensor);
  ASSERT_EQ(errors::OK, s.code()) << s.error_message();

  auto num_dims_analytical = TF_NumDims(analytical_tensor);
  ASSERT_EQ(dims.size(), num_dims_analytical);
  for (int j = 0; j < num_dims_analytical; j++) {
    auto dim_analytical = TF_Dim(analytical_tensor, j);
    ASSERT_EQ(dims[j], dim_analytical);
  }

  int64_t num_elem_analytical = TF_TensorElementCount(analytical_tensor);
  float* danalytical = new float[num_elem_analytical]{0};
  memcpy(&danalytical[0], TF_TensorData(analytical_tensor),
         TF_TensorByteSize(analytical_tensor));

  for (int j = 0; j < num_elem_analytical; j++) {
    if (abs_error == 0)
      ASSERT_EQ(manuals[j], danalytical[j]);
    else
      ASSERT_NEAR(manuals[j], danalytical[j], abs_error);
  }
  TF_DeleteTensor(analytical_tensor);
  delete[] danalytical;
}

// nn_grad_test.cc
TEST_P(CppGradients, TestReluGrad) {
  float X_vals[] = {1.0f, 2.0f, 3.0f, -5.0f, -4.0f, -3.0f, 2.0f, 10.0f, -1.0f};
  int64_t X_dims[] = {3, 3};
  AbstractTensorHandlePtr X;
  {
    AbstractTensorHandle* X_raw;
    Status s =
        TestTensorHandleWithDimsFloat(ctx_.get(), X_vals, X_dims, 2, &X_raw);
    ASSERT_EQ(errors::OK, s.code()) << s.error_message();
    X.reset(X_raw);
  }

  ASSERT_NO_FATAL_FAILURE(CompareNumericalAndAutodiffGradients(
      ReluModel, ReluGradModel, ctx_.get(), {X.get()}, UseFunction()));

  // Mathematically, Relu isn't differentiable at `0`. So `gradient_checker`
  // does not work with it.
  AbstractTensorHandlePtr Y;
  {
    AbstractTensorHandle* Y_raw;
    Status s = TestScalarTensorHandle(ctx_.get(), 0.0f, &Y_raw);
    ASSERT_EQ(errors::OK, s.code()) << s.error_message();
    Y.reset(Y_raw);
  }

  std::vector<AbstractTensorHandle*> outputs(1);
  auto s = RunModel(ReluGradModel, ctx_.get(), {Y.get()}, absl::MakeSpan(outputs),
                    /*use_function=*/UseFunction());
  ASSERT_EQ(errors::OK, s.code()) << s.error_message();
  ASSERT_NO_FATAL_FAILURE(CheckTensorValue(
      outputs[0], {0.0f}, {{}}, UseFunction(),
      /*abs_error*/ 0));
  outputs[0]->Unref();
}
```

Reply:
> What do you think about something like

I think it is easier to follow indeed. But IMo there is one downside with this approach, we have to manually call `RunModel`, loop though the output gradients. We can not hide the implementation and we can not control how developers write tests. What do you think ?

Reply:
> I think it is easier to follow indeed. But IMO there is one downside with this approach, we have to manually call RunModel, loop though the output gradients. We can not hide the implementation and we can not control how developers write tests. What do you think ?

I think we should favor simplicity and readability over conciseness. If we think this is becoming a common enough pattern we can refactor more logic out but for now I think calling `RunModel` and checking each output manually should be ok.

Reply:
Done